libc/string/string_tokpbrkstr.c

```diff
@@ -0,0 +1,687 @@
+
+	/*
+	 * After passing all delimiters we cannot avoid misplaced tokens because
+	 * strtok looks for all elements in "ImDr" and breaks a string
+	 * in points where the first occurrence of it is found
+	 */
+	TEST_ASSERT_EQUAL_PTR(loremStr, strtok_r(loremStr, sep, &rest));
+	TEST_ASSERT_EQUAL_PTR(&loremStr[3], strtok_r(NULL, sep, &rest));
+	TEST_ASSERT_EQUAL_PTR(&loremStr[5], strtok_r(NULL, sep, &rest));
+	TEST_ASSERT_EQUAL_PTR(&loremStr[7], strtok_r(NULL, sep, &rest));
+	TEST_ASSERT_EQUAL_PTR(&loremStr[11], strtok_r(NULL, sep, &rest));
+	TEST_ASSERT_EQUAL_PTR(&loremStr[13], strtok_r(NULL, sep, &rest));
+	TEST_ASSERT_EQUAL_PTR(NULL, strtok_r(NULL, sep, &rest));
+
+
+	/*
+	 * In this case, we tokenize the whole ASCII set using jump-over by
+	 * 2 elements. We are forced to jump over 2 places because strtok set
+	 *  divider placed as a null term and in this case, we are unable to
+	 * cover the whole set. Jumping over 2 places guarantees us output in the form
+	 * of one element form ascii set to do an assert with the place where
+	 * strtok has saved the stopping place
+	 */
+	TEST_ASSERT_EQUAL_PTR(&asciiStr[1], strtok_r(&asciiStr[1], "\2", &rest));
+	for (i = 4; i < ASCII_LENGTH; i += 2) {
+		asciiSep[0] = i;
+		TEST_ASSERT_EQUAL_PTR(&asciiStr[i - 1], strtok_r(NULL, &asciiSep[0], &rest));
+	}
+
+	/* Checking if set is empty */
+	TEST_ASSERT_EQUAL_PTR(NULL, strtok_r(NULL, &asciiSep[0], &rest));
+
+	free(asciiStr);
+#endif
+}
+
+
+TEST(string_tok_r, out_of_ascii)
+{
+#ifdef __phoenix__
+	TEST_IGNORE();
+#else
+
+	char divider[6] = { 0 },
+		 separator[ASCII_LENGTH],
+		 *rest,
+		 *extAsciiStr = create_extAscii_set();
+
+	int i;
+
+	divider[0] = extAsciiStr[0];
+	divider[1] = extAsciiStr[64];
+	divider[2] = extAsciiStr[128];
+	divider[3] = extAsciiStr[192];
+	divider[4] = extAsciiStr[254];
+
+	/* Checking ability for tokenize element above standard charset */
+	TEST_ASSERT_EQUAL_PTR(&extAsciiStr[1], strtok_r(extAsciiStr, divider, &rest));
+	TEST_ASSERT_EQUAL_PTR(&extAsciiStr[65], strtok_r(NULL, divider, &rest));
+	TEST_ASSERT_EQUAL_PTR(&extAsciiStr[129], strtok_r(NULL, divider, &rest));
+	TEST_ASSERT_EQUAL_PTR(&extAsciiStr[193], strtok_r(NULL, divider, &rest));
+
+	/* 256 is not taken as a token because on element after it is a null term and it is recognized as the end of set */
+	TEST_ASSERT_EQUAL_PTR(NULL, strtok_r(NULL, divider, &rest));
+
+	free(extAsciiStr);
+	extAsciiStr = create_extAscii_set();
+	/*
+	 * Checking available tokenization on all elements
+	 * from outside of ASCII charset one by one, with
+	 * the first separator set as the first element of the set
+	 */
+	TEST_ASSERT_EQUAL_PTR(extAsciiStr, strtok_r(extAsciiStr, "\128", &rest));
+	for (i = ASCII_LENGTH + 1; i < sizeof(separator); i++) {
+		separator[i - ASCII_LENGTH] = i;
+		TEST_ASSERT_EQUAL_PTR(&extAsciiStr[i - ASCII_LENGTH], strtok_r(NULL, &separator[i - ASCII_LENGTH], &rest));
+	}
+
+	free(extAsciiStr);
+#endif
+}
+
+
+TEST(string_tok_r, restriction)
```

Comment:
What is the intention of this test? Please leave a comment (the test case name does not show it sufficiently)

Reply:
This test case was created to check state restrictions about managing which tokenization has a place at this particular moment.

Reply:
I don't quite understand, first call puts a NULL pointer in asciiStr[2], so each of next calls will return only "1" as a token, I don't see a purpose of asserting it 255 times.

I suppose you wanted to test mainly the rest there, which in fact is the same for all iterations (I don't see why only two of the last rests have to be equal)